### PR TITLE
functoriality of binary coproducts

### PR DESCRIPTION
**Pull request by Alizter**
<s>This is a formal dual of the proofs for the product. I haven't been able to get the proofs to follow definitionally from the functoriality of the products however. Opping in the presence of functors is something I find very confusing so to be more practical I just reproved it. </s>

Using lemmas about functors on opposite categories we can derive the functionality of coproducts using the functoriality of products.

### CODE REVIEW

**Comment by Alizter** [comment]
The issue comes from needing the "opposite" of a statement like:
```
Global Instance is0functor_op A B (F : A -> B)
  `{IsGraph A, IsGraph B, x : !Is0Functor F}
  : Is0Functor (F : A^op -> B^op).
```
however the lack of control on the instances makes this difficult to do.

**Comment by jdchristensen** [comment]
I think we need two helper lemmas.  The first says that if `F : A -> A` is a 0-functor, then `F : op A -> op A` is a 0-functor.  And then a corresponding one for 1-functors.  These will likely be useful in many places where dualization is happening.

IIUC, these can't be definitional, but the lemmas will be easy to prove.

**Comment by jdchristensen** [comment]
Our messages crossed.  What you are proposing is what I was suggesting, except that it won't be proven by instance search, but by using `Build_Is0Functor` and supplying the data.

**Comment by jdchristensen** [comment]
Oh, sorry, now I see that the helper lemmas I'm asking for are already there.  Let me look at this again.

**Comment by Alizter** [comment]
I originally tried that but should A^op have its own 1-cat instance or the ones from being an opposite category?

**Comment by jdchristensen** [comment]
This works.
```coq
Global Instance is0functor_op' A B (F : A^op -> B^op)
  `{IsGraph A, IsGraph B, Fop : !Is0Functor (F : A^op -> B^op)}
  : Is0Functor (F : A -> B).
Proof.
  apply Build_Is0Functor.
  intros a b f.
  exact (@fmap A^op B^op _ _ F Fop b a f).
Defined.

Global Instance is0functor_cat_coprod_l {A : Type}
  `{HasBinaryCoproducts A} y
  : Is0Functor (A:=A) (fun x => cat_coprod x y).
Proof.
  Set Printing Implicit.
  rapply is0functor_op'.
  exact (is0functor_cat_prod_l (A:=A^op) (H0:=H0) y).
Defined.
```
The first one should be in Opposites.v, of course.

**Comment by Alizter** [comment]
@jdchristensen Nice! I messed around with that for hours getting nowhere before, thanks alot!

File: theories/WildCat/Coproducts.v [hunk 1]
```diff
@@ -142,6 +142,38 @@ Defined.
 
 (** TODO: The functoriality argument doesn't follow definitionally from the functoriality of [cat_prod], however after some modification it is close. We need to use appropriate lemmas for opposite functors. *) 
 
+Global Instance is0functor_cat_coprod_l {A : Type}
+  `{HasBinaryCoproducts A} y
+  : Is0Functor (A:=A) (fun x => cat_coprod x y).
+Proof.
+  rapply is0functor_op'.
+  exact (is0functor_cat_prod_l (A:=A^op) (H0:=H0) y).
```
[the remark this answers is not in the document]
Comment (Alizter) [reply theories/WildCat/Coproducts.v:150]: I'll add a comment about using `Set Printing Implicit`.

**Pull request merged**